libnmstate/nm/dns.py

```diff
@@ -0,0 +1,101 @@
+#
+# Copyright 2019 Red Hat, Inc.
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 2 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <https://www.gnu.org/licenses/>.
+#
+from operator import itemgetter
+from operator import methodcaller
+
+from libnmstate import iplib
+from libnmstate.nm import nmclient
+from libnmstate.schema import DNS
+from libnmstate.nm import ipv4 as nm_ipv4
+from libnmstate.nm import ipv6 as nm_ipv6
+from libnmstate.error import NmstateInternalError
+
+
+DNS_DEFAULT_PRIORITY_VPN = 50
+DNS_DEFAULT_PRIORITY_OTHER = 100
+
+IPV6_ADDRESS_LENGTH = 128
+
+
+def get_running():
+    dns_state = {
+        DNS.SERVER: [],
+        DNS.SEARCH: []
+    }
+    client = nmclient.client()
+    dns_confs = client.get_dns_configuration()
+    for dns_conf in sorted(dns_confs, key=methodcaller('get_priority')):
+        iface_name = dns_conf.get_interface()
+        for ns in dns_conf.get_nameservers():
+            if iplib.is_ipv6_link_local_addr(ns, IPV6_ADDRESS_LENGTH):
+                if not iface_name:
+                    # For IPv6 link local address, the interface name should be
+                    # appended also.
+                    raise NmstateInternalError(
+                        "Got empty interface name for IPv6 link local "
+                        "DNS nameserver entry {}".format(ns))
+                ns_addr = "{}%{}".format(ns, iface_name)
+            else:
+                ns_addr = ns
+            dns_state[DNS.SERVER].append(ns_addr)
+        dns_state[DNS.SEARCH].extend(dns_conf.get_domains())
+    if not dns_state[DNS.SERVER] and not dns_state[DNS.SEARCH]:
+        return {}
+    return dns_state
+
+
+def get_config():
+    dns_conf = {
+        DNS.SERVER: [],
+        DNS.SEARCH: []
+    }
+    tmp_dns_confs = []
+    client = nmclient.client()
+    for ac in client.get_active_connections():
+        # NM prefers IPv6 over IPv4 DNS.
+        for ip_profile in (nm_ipv6.get_ip_profile(ac),
+                           nm_ipv4.get_ip_profile(ac)):
+            if not ip_profile:
+                continue
+            if not ip_profile.props.dns and not ip_profile.props.dns_search:
+                continue
+            priority = ip_profile.props.dns_priority
+            if priority == 0:
```

Comment:
the nm.settings documentation mention that there can also be a globally configured default priority that would be used if `0` is used here. Does this need to be handled somehow?

Reply:
It was controlled by `NetworkManager.conf` where libnm does not provide any function to query/set it. The same thing also applied to many other options.
I guess we will have to see how feedback comes.

Reply:
ok, should we document this?

Reply:
maybe in a comment?